webhooks/clusterclass.go

```diff
@@ -161,23 +160,25 @@ func (webhook *ClusterClass) validateCompatibleSpecChanges(old, in *clusterv1.Cl
 	allErrs = append(allErrs, webhook.validateMachineDeploymentsCompatibleChanges(old, in)...)
 
 	// Validate InfrastructureClusterTemplate changes in a compatible way.
-	allErrs = append(allErrs, webhook.validateTemplatesAreCompatible(in.Spec.Infrastructure,
-		old.Spec.Infrastructure,
-		field.NewPath("spec", "infrastructure"),
-	)...)
+	if err := check.TemplatesAreCompatible(in.Spec.Infrastructure, old.Spec.Infrastructure); err != nil {
+		allErrs = append(allErrs, field.Invalid(field.NewPath("spec", "infrastructure"),
+			in.Spec.Infrastructure, err.Error()),
+		)
+	}
 
 	// Validate control plane changes in a compatible way.
-	allErrs = append(allErrs, webhook.validateTemplatesAreCompatible(in.Spec.ControlPlane.LocalObjectTemplate,
-		old.Spec.ControlPlane.LocalObjectTemplate,
-		field.NewPath("spec", "controlPlane"),
-	)...)
+	if err := check.TemplatesAreCompatible(in.Spec.ControlPlane.LocalObjectTemplate, old.Spec.ControlPlane.LocalObjectTemplate); err != nil {
+		allErrs = append(allErrs, field.Invalid(field.NewPath("spec", "controlPlane", "localObjectTemplate"),
```

Comment:
```suggestion
		allErrs = append(allErrs, field.Invalid(field.NewPath("spec", "controlPlane"),
```

I think we shouldn't add `localObjectTemplate` here as that's not in the user facing JSON path. (only in the go structs)

I think we should instead append "ref" to all those paths here. Because the error will be related to `.spec.controlPlane.ref` (similar in other cases)

Reply:
This is changed with the newer template checks using field.ErrorList I think.